redfish-core/lib/metric_report_definition.hpp

```diff
@@ -200,30 +257,91 @@ inline bool getUserParameters(crow::Response& res, const crow::Request& req,
+}
+inline bool toDbusMetrics(crow::Response& res,
+                          std::vector<nlohmann::json>& metrics,
+                          AddReportArgs& args)
+{
+    uint16_t itemNb = 0;
     args.metrics.reserve(metrics.size());
     for (auto& m : metrics)
     {
         std::string id;
         std::vector<std::string> uris;
+        std::optional<std::string> function;
+        std::optional<std::string> scope;
+        std::optional<std::string> durationStr;
         if (!json_util::readJson(m, res, "MetricId", id, "MetricProperties",
-                                 uris))
+                                 uris, "CollectionFunction", function,
+                                 "CollectionTimeScope", scope,
+                                 "CollectionDuration", durationStr))
         {
             return false;
         }
 
-        args.metrics.emplace_back(std::move(id), std::move(uris));
+        function = function.has_value() ? function : "Average";
+        scope = scope.has_value() ? scope : "Point";
+        durationStr = durationStr.has_value() ? durationStr : "PT0S";
+
+        std::string operation = getOperationType(*function, MethodType::post);
+        if (operation.empty())
+        {
+            messages::propertyValueNotInList(
+                res, *function, "CollectionFunction/" + std::to_string(itemNb));
+            return false;
+        }
+        if (!isTimeScopeValid(*scope))
+        {
+            messages::propertyValueNotInList(
+                res, *scope, "CollectionTimeScope/" + std::to_string(itemNb));
+            return false;
+        }
+        std::optional<std::chrono::milliseconds> durationNum =
+            time_utils::fromDurationString(*durationStr);
+        if (!durationNum)
+        {
+            messages::propertyValueIncorrect(res, "CollectionDuration",
+                                             *durationStr);
+            return false;
+        }
+        args.metrics.emplace_back(std::move(id), std::move(uris),
+                                  std::move(operation), std::move(*scope),
+                                  static_cast<uint64_t>(durationNum->count()));
+        itemNb++;
+    }
+
+    return true;
+}
+
+inline bool getUserParameters(crow::Response& res, const crow::Request& req,
+                              AddReportArgs& args)
+{
+    std::vector<nlohmann::json> metrics;
+    std::vector<std::string> reportActions;
+    std::optional<nlohmann::json> schedule;
+    if (!json_util::readJson(req, res, "Id", args.name, "Metrics", metrics,
+                             "MetricReportDefinitionType", args.reportingType,
+                             "ReportActions", reportActions, "Schedule",
+                             schedule) ||
+        !isIdValid(res, args.name) ||
+        !toDbusReportActions(res, reportActions, args) ||
+        !toDbusReportingType(res, schedule, args) ||
+        !toDbusMetrics(res, metrics, args))
+    {
+        return false;
```

Comment:
This should be split up at each OR condition.

Also, there should be an BMCWEB_LOG_ERROR message in each of these isIdValid, toDbusReportActions, toDbusReportingTypes, toDbusMetrics describing which called failed.